<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.8.1" />
<title>autostreamtree.params API documentation</title>
<meta name="description" content="" />
<link href='https://cdnjs.cloudflare.com/ajax/libs/normalize/8.0.0/normalize.min.css' rel='stylesheet'>
<link href='https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/8.0.0/sanitize.min.css' rel='stylesheet'>
<link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" rel="stylesheet">
<style>.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>autostreamtree.params</code></h1>
</header>
<section id="section-intro">
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">import sys
import getopt


# Object to parse command-line arguments
class parseArgs():
    def __init__(self):
        # Define options
        try:
            options, _ = getopt.getopt(
                sys.argv[1:], &#39;hs:i:r:p:d:a:lw:o:gP:L:Scn:G:v:C:O:&#39;,
                [&#34;shp=&#34;, &#34;help&#34;, &#34;input=&#34;, &#34;run=&#34;, &#34;pop=&#34;, &#34;popmap=&#34;, &#34;dist=&#34;,
                 &#34;agg_method=&#34;, &#34;het&#34;, &#34;genmat=&#34;, &#34;snp&#34;, &#34;snps&#34;, &#34;msat&#34;,
                 &#34;msats&#34;, &#34;log&#34;, &#34;and_log&#34;, &#34;iterative&#34;, &#34;weight=&#34;, &#34;out=&#34;,
                 &#34;method=&#34;, &#34;plots&#34;, &#34;plot&#34;, &#34;perm=&#34;, &#34;phased&#34;, &#34;median&#34;,
                 &#34;diploid&#34;, &#34;geopop&#34;, &#34;geopops&#34;, &#34;global_het&#34;, &#34;haploid&#34;,
                 &#34;loc_agg=&#34;, &#34;pop_agg=&#34;, &#34;sdist_agg=&#34;, &#34;clusterpop&#34;,
                 &#34;epsilon=&#34;, &#34;min_samples=&#34;, &#34;sclusterpop&#34;, &#34;network=&#34;,
                 &#34;overwrite&#34;, &#34;reachid_col=&#34;, &#34;length_col=&#34;, &#34;coercemat&#34;,
                 &#34;locmatdir=&#34;, &#34;vcf=&#34;, &#34;concat=&#34;, &#34;edge_list=&#34;, &#34;gdf_out=&#34;,
                 &#34;seed=&#34;]
            )
        except getopt.GetoptError as err:
            print(err)
            self.display_help(
                &#34;\nExiting because getopt returned non-zero exit status.&#34;)
        # Default values for params
        # Input params
        self.shapefile = None
        self.locmatdir = None
        self.geodb = None
        self.vcf = None
        self.concat = &#34;none&#34;
        self.run = &#34;ALL&#34;
        self.network = None
        self.pop = False
        self.overwrite = False
        self.geopop = False
        self.clusterpop = False
        self.sclusterpop = False
        self.output_driver = &#34;GPKG&#34;
        self.dist = &#34;LINFST&#34;
        self.het = False
        self.genmat = None
        self.log = False
        self.and_log = False
        self.iterative = True
        self.weight = &#34;CSE67&#34;
        self.permutations = 1000
        self.coercemat = False
        self.method = &#34;PEARSON&#34;
        self.edge_list = None
        self.plots = False
        self.out = &#34;out&#34;
        self.median = False
        self.ploidy = 2
        self.global_het = False
        self.loc_agg = &#34;ARITH&#34;
        self.pop_agg = &#34;ARITH&#34;
        self.sdist_agg = &#34;ARITH&#34;
        self.seed = None

        self.reachid_col = &#34;HYRIV_ID&#34;
        self.length_col = &#34;LENGTH_KM&#34;

        # dbscan Options
        self.min_samples = 1
        self.epsilon = 20

        # First pass to see if help menu was called
        for o, a in options:
            if o in (&#34;-h&#34;, &#34;-help&#34;, &#34;--help&#34;):
                self.display_help(&#34;Exiting because help menu was called.&#34;)

        # Second pass to set all args.
        for opt, arg in options:
            arg = arg.strip()
            opt = opt.replace(&#34;-&#34;, &#34;&#34;)

            if opt in (&#39;s&#39;, &#39;shp&#39;):
                self.shapefile = arg
            elif opt in (&#39;h&#39;, &#39;help&#39;):
                pass
            elif opt in (&#39;i&#39;, &#39;coords&#39;):
                self.geodb = arg
            elif opt in (&#34;C&#34;, &#34;concat&#34;):
                arg_lower = str(arg).lower()
                if arg_lower in (&#34;none&#34;, &#34;all&#34;, &#34;loc&#34;):
                    self.concat = arg_lower
                else:
                    self.display_help(f&#34;Invalid option {arg_lower} \
                                      for option &lt;--concat&gt;&#34;)
            elif opt in (&#39;v&#39;, &#39;vcf&#39;):
                self.vcf = arg
            elif opt == &#34;seed&#34;:
                self.seed = int(arg)
            elif opt in (&#39;r&#39;, &#39;run&#39;):
                self.run = arg.upper()
                valid_runs = [&#34;ALL&#34;, &#34;GENDIST&#34;, &#34;IBD&#34;, &#34;STREAMDIST&#34;,
                              &#34;STREAMTREE&#34;, &#34;DISTANCES&#34;, &#34;RUNLOCI&#34;]
                if self.run not in valid_runs:
                    self.display_help(f&#34;Invalid option {self.run} \
                                      for option &lt;-r/--run&gt;&#34;)
            elif opt in (&#39;p&#39;, &#39;pop&#39;, &#34;popmap&#34;):
                self.pop = arg
            elif opt in (&#34;g&#34;, &#34;geopop&#34;, &#34;geopops&#34;):
                self.geopop = True
            elif opt in (&#39;d&#39;, &#39;dist&#39;):
                self.dist = arg.upper()
                valid_dists = [&#34;PDIST&#34;, &#34;FST&#34;,
                               &#34;LINFST&#34;, &#34;JOST&#34;, &#34;NEI83&#34;,
                               &#34;CHORD&#34;, &#34;GST&#34;, &#34;HARMD&#34;]
                if self.dist not in valid_dists:
                    self.display_help(f&#34;Invalid option {self.dist} for \
                                      option &lt;-d/--dist&gt;&#34;)
            elif opt == &#34;het&#34;:
                self.het = True
            elif opt == &#34;locmatdir&#34;:
                self.locmatdir = arg
            elif opt == &#34;reachid_col&#34;:
                self.reachid_col = arg
            elif opt == &#34;length_col&#34;:
                self.length_col = arg
            elif opt in (&#34;genmat&#34;, &#34;G&#34;):
                self.genmat = arg
            elif opt in (&#39;l&#39;, &#39;log&#39;):
                self.log = True
            elif opt == &#34;and_log&#34;:
                self.and_log = True
            elif opt == &#34;edge_list&#34;:
                self.edge_list = arg
            elif opt in (&#34;clusterpop&#34;, &#34;c&#34;):
                self.clusterpop = True
            elif opt == &#34;epsilon&#34;:
                self.epsilon = float(arg)
            elif opt == &#34;min_samples&#34;:
                self.min_samples = int(arg)
            elif opt in (&#39;w&#39;, &#39;weight&#39;):
                self.weight = arg.upper()
                weight_mappings = {
                    &#34;FM&#34;: &#34;FM67&#34;, &#34;1/D&#34;: &#34;FM67&#34;,
                    &#34;BEYER&#34;: &#34;BEYER74&#34;, &#34;1/D^2&#34;: &#34;BEYER74&#34;,
                    &#34;1&#34;: &#34;CSE67&#34;, &#34;CSE&#34;: &#34;CSE67&#34;
                }
                self.weight = weight_mappings.get(self.weight, self.weight)
                valid_weights = [&#34;CSE67&#34;, &#34;FM67&#34;, &#34;BEYER74&#34;]
                if self.weight not in valid_weights:
                    self.display_help(f&#34;Invalid option {self.weight} \
                                      for option &lt;-w/--weight&gt;&#34;)
            elif opt in (&#39;o&#39;, &#39;out&#39;):
                self.out = arg
            elif opt == &#34;stream_fit&#34;:
                self.stream_fit = True
            elif opt == &#34;overwrite&#34;:
                self.overwrite = True
            elif opt == &#34;perm&#34;:
                self.permutations = int(arg)
            elif opt == &#34;method&#34;:
                print(&#34;Sorry: Option --method is not yet implemented.&#34;)
                sys.exit(0)
            elif opt in (&#34;plot&#34;, &#34;plots&#34;):
                self.plots = True
            elif opt == &#34;phased&#34;:
                self.phased = True
                print(&#34;WARNING: Option &lt;--snp&gt; not yet implemented&#34;)
            elif opt == &#34;median&#34;:
                self.median = True
            elif opt == &#34;diploid&#34;:
                self.ploidy = 2
            elif opt == &#34;haploid&#34;:
                self.ploidy = 1
            elif opt in (&#34;O&#34;, &#34;gdf_out&#34;):
                arg_upper = str(arg).upper()
                if arg_upper not in [&#34;GPKG&#34;, &#34;SHP&#34;, &#34;GDB&#34;]:
                    self.display_help(f&#34;Invalid option {arg_upper} for \
                                      option &lt;--gdf_out&gt;&#34;)
                self.output_driver = arg_upper
            elif opt == &#34;global_het&#34;:
                self.global_het = True
            elif opt == &#34;coercemat&#34;:
                self.coercemat = True
            elif opt in (&#34;network&#34;, &#34;n&#34;):
                self.network = arg
            elif opt in (&#34;pop_agg&#34;, &#34;P&#34;):
                self.pop_agg = arg.upper()
                valid_agg_methods = [&#34;HARM&#34;, &#34;ADJHARM&#34;, &#34;ARITH&#34;,
                                     &#34;GEOM&#34;, &#34;MEDIAN&#34;, &#34;MAX&#34;, &#34;MIN&#34;]
                if self.pop_agg not in valid_agg_methods:
                    self.display_help(f&#34;Invalid option {self.pop_agg} for \
                                      option &lt;--pop_agg&gt;&#34;)
            elif opt in (&#34;loc_agg&#34;, &#34;L&#34;):
                self.loc_agg = arg.upper()
                valid_agg_methods = [&#34;HARM&#34;, &#34;ADJHARM&#34;, &#34;ARITH&#34;, &#34;GEOM&#34;,
                                     &#34;MEDIAN&#34;, &#34;MAX&#34;, &#34;MIN&#34;]
                if self.loc_agg not in valid_agg_methods:
                    self.display_help(f&#34;Invalid option {self.loc_agg} for \
                                      option &lt;--loc_agg&gt;&#34;)
            else:
                assert False, f&#34;Unhandled option {opt}&#34;

        if not self.geodb:
            self.display_help(&#34;No input provided &lt;-i/--coords&gt;&#34;)
        if not self.shapefile and self.run != &#34;GENDIST&#34;:
            self.display_help(&#34;No shapefile provided &lt;-s/--shp&gt;&#34;)
        if self.ploidy &gt; 2 or self.ploidy &lt; 1:
            self.display_help(
                f&#34;Ploidy of {self.ploidy} not currently \
                    allowable. Please choose 1 (haploid) or 2 (diploid)&#34;)

        # sanity checks
        if self.dist not in [&#34;PDIST&#34;]:
            if not self.pop and not self.geopop:
                self.display_help(
                    f&#34;ERROR: Distance metric {self.dist} not possible \
                        without --pop or --geopop data.&#34;
                )

    def display_help(self, message=None):
        if message is not None:
            print(&#34;\n&#34;, message)

        print(&#34;\nautostreamtree\n&#34;)
        print(&#34;Author: Tyler K Chafin, Biomathematics and Statistics Scotland&#34;)
        print(&#34;Description: Methods for analysing genetic distances in \
networks.&#34;)
        print(
            &#34;\nMandatory arguments:\n&#34;
            &#34;    -s, --shp       : Path to shapefile containing cleaned, \
contiguous stream reaches\n&#34;
            &#34;                       (can also support geodatabase or GPKG \
files)\n&#34;
            &#34;    -i, --input     : Input .tsv file containing sample \
coordinates\n&#34;
            &#34;    -v, --vcf       : Input VCF file containing genotypes\n\n&#34;

            &#34;General options:\n&#34;
            &#34;    -o, --out       : Output prefix [default=\&#34;out\&#34;]\n&#34;
            &#34;    -O, --gdf_out   : Output driver for annotated geodataframe \
(options \&#34;SHP\&#34;, \&#34;GPKG\&#34;, \&#34;GDB\&#34;)\n&#34;
            &#34;    -C, --concat    : Concatenate all SNPs (\&#34;all\&#34;), by locus \
(\&#34;loc\&#34;), or not at all (\&#34;none\&#34;)\n&#34;
            &#34;    -n, --network   : Provide an already optimized network \
output from a previous run\n&#34;
            &#34;    --overwrite     : Overwrite an input network (Only relevant \
with --network)\n&#34;
            &#34;    -h, --help      : Displays help menu\n&#34;
            &#34;    -r, --run       : Run which steps? Options:\n&#34;
            &#34;        ALL         : Run all steps\n&#34;
            &#34;        GENDIST     : Only calculate genetic distance matrix\n&#34;
            &#34;        STREAMDIST  : Only compute pairwise stream distances\n&#34;
            &#34;        DISTANCES   : Only compute GENDIST + STREAMDIST\n&#34;
            &#34;        IBD         : GENDIST + STREAMDIST + Mantel test\n&#34;
            &#34;        STREAMTREE  : GENDIST + STREAMDIST + fit StreamTree \
model\n&#34;
            &#34;        RUNLOCI     : Run STREAMTREE fitting on each locus\n&#34;
            &#34;    -p, --pop       : Pool individuals based on an input \
population map tsv file\n&#34;
            &#34;        NOTE: The location will be taken as the centroid among \
individual samples\n&#34;
            &#34;    -g, --geopop    : Pool individuals having identical \
coordinates\n&#34;
            &#34;    -c, --clusterpop: Use DBSCAN algorithm to automatically \
cluster populations\n&#34;
            &#34;    --reachid_col   : Attribute name representing primary key in \
shapefile [default=\&#34;HYRIV_ID\&#34;]\n&#34;
            &#34;    --length_col    : Attribute name giving length in kilometers \
[default=\&#34;LENGTH_KM\&#34;]\n\n&#34;
            &#34;    --seed          : Seed for RNG\n\n&#34;

            &#34;Genetic distance options:\n&#34;
            &#34;    -d, --dist      : Use which metric of distance? Options:\n&#34;
            &#34;        Individual-based:\n&#34;
            &#34;          PDIST     : Uncorrected p-distances [# Differences / \
Length]\n&#34;
            &#34;        Frequency models (when using --pop):\n&#34;
            &#34;          FST       : Weir and Cockerham&#39;s Fst formulation \
(=THETAst)\n&#34;
            &#34;          LINFST    : [default] Rousset&#39;s (1997) Fst \
[=Fst/(1-Fst)]\n&#34;
            &#34;          JOST      : Global estimate Jost&#39;s (2008) D\n&#34;
            &#34;          HARMD     : Harmonic mean of Jost&#39;s D per-locus\n&#34;
            &#34;          CHORD     : Cavalli-Sforza and Edwards (1967) chord \
distance\n&#34;
            &#34;          --NOTE: Individual-based metrics can also be computed \
for\n&#34;
            &#34;                    populations. You can set how these are \
aggregated w/ --pop_agg\n&#34;
            &#34;          --NOTE: Multiple loci for PDIST\n&#34;
            &#34;                  will be reported using the method defined in \
--loc_agg\n&#34;
            &#34;    -G, --genmat    : Skip calculation and use the provided \
labeled .tsv matrix\n&#34;
            &#34;    --coercemat     : [Boolean] Coerce negative values in input \
matrix to zero\n&#34;
            &#34;    --het           : [Boolean] Count partial differences [e.g. \
ind1=T, ind2=W]\n&#34;
            &#34;    --global_het    : Estimate Ht using global frequencies \
(default is averaged over pops)\n\n&#34;

            &#34;DBSCAN options (only when --clusterpop):\n&#34;
            &#34;    --min_samples   : Minimum samples per cluster [default=1]\n&#34;
            &#34;    --epsilon       : Maximum distance (in km) within a cluster \
[default=20]\n\n&#34;

            &#34;Aggregation options:\n&#34;
            &#34;    -P, --pop_agg   : Define aggregator function for certain \
genetic distances in pop samples\n&#34;
            &#34;    -L, --loc_agg   : Define aggregator function for aggregating \
locus-wise distances\n&#34;
            &#34;        All of these can take the following options:\n&#34;
            &#34;          ARITH     : [default] Use arithmetic mean\n&#34;
            &#34;          MEDIAN    : Use median distance\n&#34;
            &#34;          HARM      : Use harmonic mean\n&#34;
            &#34;          ADJHARM   : Adjusted harmonic mean (see docs)\n&#34;
            &#34;          GEOM      : Use geometric mean\n&#34;
            &#34;          MIN       : Use minimum distance\n&#34;
            &#34;          MAX       : Use maximum distance\n\n&#34;

            &#34;IBD options:\n&#34;
            &#34;    --perm          : Number of permutations for mantel test \
[def=1000]\n&#34;
            &#34;    --and_log       : Also perform IBD steps with log geographic \
distances\n\n&#34;

            &#34;StreamTree options (see Kalinowski et al. 2008) :\n&#34;
            &#34;    -w, --weight    : Desired weighting for least-squares \
fitting:\n&#34;
            &#34;        Options:\n&#34;
            &#34;          FM67      : Fitch and Margoliash (1967) [w = 1/D^2]\n&#34;
            &#34;          BEYER74   : Beyer et al. (1974) weights [w = 1/D]\n&#34;
            &#34;          CSE67     : [default] Cavalli-Sforza &amp; Edwards (1967) \
[w = 1]\n\n&#34;
        )
        print()
        sys.exit()</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="autostreamtree.params.parseArgs"><code class="flex name class">
<span>class <span class="ident">parseArgs</span></span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class parseArgs():
    def __init__(self):
        # Define options
        try:
            options, _ = getopt.getopt(
                sys.argv[1:], &#39;hs:i:r:p:d:a:lw:o:gP:L:Scn:G:v:C:O:&#39;,
                [&#34;shp=&#34;, &#34;help&#34;, &#34;input=&#34;, &#34;run=&#34;, &#34;pop=&#34;, &#34;popmap=&#34;, &#34;dist=&#34;,
                 &#34;agg_method=&#34;, &#34;het&#34;, &#34;genmat=&#34;, &#34;snp&#34;, &#34;snps&#34;, &#34;msat&#34;,
                 &#34;msats&#34;, &#34;log&#34;, &#34;and_log&#34;, &#34;iterative&#34;, &#34;weight=&#34;, &#34;out=&#34;,
                 &#34;method=&#34;, &#34;plots&#34;, &#34;plot&#34;, &#34;perm=&#34;, &#34;phased&#34;, &#34;median&#34;,
                 &#34;diploid&#34;, &#34;geopop&#34;, &#34;geopops&#34;, &#34;global_het&#34;, &#34;haploid&#34;,
                 &#34;loc_agg=&#34;, &#34;pop_agg=&#34;, &#34;sdist_agg=&#34;, &#34;clusterpop&#34;,
                 &#34;epsilon=&#34;, &#34;min_samples=&#34;, &#34;sclusterpop&#34;, &#34;network=&#34;,
                 &#34;overwrite&#34;, &#34;reachid_col=&#34;, &#34;length_col=&#34;, &#34;coercemat&#34;,
                 &#34;locmatdir=&#34;, &#34;vcf=&#34;, &#34;concat=&#34;, &#34;edge_list=&#34;, &#34;gdf_out=&#34;,
                 &#34;seed=&#34;]
            )
        except getopt.GetoptError as err:
            print(err)
            self.display_help(
                &#34;\nExiting because getopt returned non-zero exit status.&#34;)
        # Default values for params
        # Input params
        self.shapefile = None
        self.locmatdir = None
        self.geodb = None
        self.vcf = None
        self.concat = &#34;none&#34;
        self.run = &#34;ALL&#34;
        self.network = None
        self.pop = False
        self.overwrite = False
        self.geopop = False
        self.clusterpop = False
        self.sclusterpop = False
        self.output_driver = &#34;GPKG&#34;
        self.dist = &#34;LINFST&#34;
        self.het = False
        self.genmat = None
        self.log = False
        self.and_log = False
        self.iterative = True
        self.weight = &#34;CSE67&#34;
        self.permutations = 1000
        self.coercemat = False
        self.method = &#34;PEARSON&#34;
        self.edge_list = None
        self.plots = False
        self.out = &#34;out&#34;
        self.median = False
        self.ploidy = 2
        self.global_het = False
        self.loc_agg = &#34;ARITH&#34;
        self.pop_agg = &#34;ARITH&#34;
        self.sdist_agg = &#34;ARITH&#34;
        self.seed = None

        self.reachid_col = &#34;HYRIV_ID&#34;
        self.length_col = &#34;LENGTH_KM&#34;

        # dbscan Options
        self.min_samples = 1
        self.epsilon = 20

        # First pass to see if help menu was called
        for o, a in options:
            if o in (&#34;-h&#34;, &#34;-help&#34;, &#34;--help&#34;):
                self.display_help(&#34;Exiting because help menu was called.&#34;)

        # Second pass to set all args.
        for opt, arg in options:
            arg = arg.strip()
            opt = opt.replace(&#34;-&#34;, &#34;&#34;)

            if opt in (&#39;s&#39;, &#39;shp&#39;):
                self.shapefile = arg
            elif opt in (&#39;h&#39;, &#39;help&#39;):
                pass
            elif opt in (&#39;i&#39;, &#39;coords&#39;):
                self.geodb = arg
            elif opt in (&#34;C&#34;, &#34;concat&#34;):
                arg_lower = str(arg).lower()
                if arg_lower in (&#34;none&#34;, &#34;all&#34;, &#34;loc&#34;):
                    self.concat = arg_lower
                else:
                    self.display_help(f&#34;Invalid option {arg_lower} \
                                      for option &lt;--concat&gt;&#34;)
            elif opt in (&#39;v&#39;, &#39;vcf&#39;):
                self.vcf = arg
            elif opt == &#34;seed&#34;:
                self.seed = int(arg)
            elif opt in (&#39;r&#39;, &#39;run&#39;):
                self.run = arg.upper()
                valid_runs = [&#34;ALL&#34;, &#34;GENDIST&#34;, &#34;IBD&#34;, &#34;STREAMDIST&#34;,
                              &#34;STREAMTREE&#34;, &#34;DISTANCES&#34;, &#34;RUNLOCI&#34;]
                if self.run not in valid_runs:
                    self.display_help(f&#34;Invalid option {self.run} \
                                      for option &lt;-r/--run&gt;&#34;)
            elif opt in (&#39;p&#39;, &#39;pop&#39;, &#34;popmap&#34;):
                self.pop = arg
            elif opt in (&#34;g&#34;, &#34;geopop&#34;, &#34;geopops&#34;):
                self.geopop = True
            elif opt in (&#39;d&#39;, &#39;dist&#39;):
                self.dist = arg.upper()
                valid_dists = [&#34;PDIST&#34;, &#34;FST&#34;,
                               &#34;LINFST&#34;, &#34;JOST&#34;, &#34;NEI83&#34;,
                               &#34;CHORD&#34;, &#34;GST&#34;, &#34;HARMD&#34;]
                if self.dist not in valid_dists:
                    self.display_help(f&#34;Invalid option {self.dist} for \
                                      option &lt;-d/--dist&gt;&#34;)
            elif opt == &#34;het&#34;:
                self.het = True
            elif opt == &#34;locmatdir&#34;:
                self.locmatdir = arg
            elif opt == &#34;reachid_col&#34;:
                self.reachid_col = arg
            elif opt == &#34;length_col&#34;:
                self.length_col = arg
            elif opt in (&#34;genmat&#34;, &#34;G&#34;):
                self.genmat = arg
            elif opt in (&#39;l&#39;, &#39;log&#39;):
                self.log = True
            elif opt == &#34;and_log&#34;:
                self.and_log = True
            elif opt == &#34;edge_list&#34;:
                self.edge_list = arg
            elif opt in (&#34;clusterpop&#34;, &#34;c&#34;):
                self.clusterpop = True
            elif opt == &#34;epsilon&#34;:
                self.epsilon = float(arg)
            elif opt == &#34;min_samples&#34;:
                self.min_samples = int(arg)
            elif opt in (&#39;w&#39;, &#39;weight&#39;):
                self.weight = arg.upper()
                weight_mappings = {
                    &#34;FM&#34;: &#34;FM67&#34;, &#34;1/D&#34;: &#34;FM67&#34;,
                    &#34;BEYER&#34;: &#34;BEYER74&#34;, &#34;1/D^2&#34;: &#34;BEYER74&#34;,
                    &#34;1&#34;: &#34;CSE67&#34;, &#34;CSE&#34;: &#34;CSE67&#34;
                }
                self.weight = weight_mappings.get(self.weight, self.weight)
                valid_weights = [&#34;CSE67&#34;, &#34;FM67&#34;, &#34;BEYER74&#34;]
                if self.weight not in valid_weights:
                    self.display_help(f&#34;Invalid option {self.weight} \
                                      for option &lt;-w/--weight&gt;&#34;)
            elif opt in (&#39;o&#39;, &#39;out&#39;):
                self.out = arg
            elif opt == &#34;stream_fit&#34;:
                self.stream_fit = True
            elif opt == &#34;overwrite&#34;:
                self.overwrite = True
            elif opt == &#34;perm&#34;:
                self.permutations = int(arg)
            elif opt == &#34;method&#34;:
                print(&#34;Sorry: Option --method is not yet implemented.&#34;)
                sys.exit(0)
            elif opt in (&#34;plot&#34;, &#34;plots&#34;):
                self.plots = True
            elif opt == &#34;phased&#34;:
                self.phased = True
                print(&#34;WARNING: Option &lt;--snp&gt; not yet implemented&#34;)
            elif opt == &#34;median&#34;:
                self.median = True
            elif opt == &#34;diploid&#34;:
                self.ploidy = 2
            elif opt == &#34;haploid&#34;:
                self.ploidy = 1
            elif opt in (&#34;O&#34;, &#34;gdf_out&#34;):
                arg_upper = str(arg).upper()
                if arg_upper not in [&#34;GPKG&#34;, &#34;SHP&#34;, &#34;GDB&#34;]:
                    self.display_help(f&#34;Invalid option {arg_upper} for \
                                      option &lt;--gdf_out&gt;&#34;)
                self.output_driver = arg_upper
            elif opt == &#34;global_het&#34;:
                self.global_het = True
            elif opt == &#34;coercemat&#34;:
                self.coercemat = True
            elif opt in (&#34;network&#34;, &#34;n&#34;):
                self.network = arg
            elif opt in (&#34;pop_agg&#34;, &#34;P&#34;):
                self.pop_agg = arg.upper()
                valid_agg_methods = [&#34;HARM&#34;, &#34;ADJHARM&#34;, &#34;ARITH&#34;,
                                     &#34;GEOM&#34;, &#34;MEDIAN&#34;, &#34;MAX&#34;, &#34;MIN&#34;]
                if self.pop_agg not in valid_agg_methods:
                    self.display_help(f&#34;Invalid option {self.pop_agg} for \
                                      option &lt;--pop_agg&gt;&#34;)
            elif opt in (&#34;loc_agg&#34;, &#34;L&#34;):
                self.loc_agg = arg.upper()
                valid_agg_methods = [&#34;HARM&#34;, &#34;ADJHARM&#34;, &#34;ARITH&#34;, &#34;GEOM&#34;,
                                     &#34;MEDIAN&#34;, &#34;MAX&#34;, &#34;MIN&#34;]
                if self.loc_agg not in valid_agg_methods:
                    self.display_help(f&#34;Invalid option {self.loc_agg} for \
                                      option &lt;--loc_agg&gt;&#34;)
            else:
                assert False, f&#34;Unhandled option {opt}&#34;

        if not self.geodb:
            self.display_help(&#34;No input provided &lt;-i/--coords&gt;&#34;)
        if not self.shapefile and self.run != &#34;GENDIST&#34;:
            self.display_help(&#34;No shapefile provided &lt;-s/--shp&gt;&#34;)
        if self.ploidy &gt; 2 or self.ploidy &lt; 1:
            self.display_help(
                f&#34;Ploidy of {self.ploidy} not currently \
                    allowable. Please choose 1 (haploid) or 2 (diploid)&#34;)

        # sanity checks
        if self.dist not in [&#34;PDIST&#34;]:
            if not self.pop and not self.geopop:
                self.display_help(
                    f&#34;ERROR: Distance metric {self.dist} not possible \
                        without --pop or --geopop data.&#34;
                )

    def display_help(self, message=None):
        if message is not None:
            print(&#34;\n&#34;, message)

        print(&#34;\nautostreamtree\n&#34;)
        print(&#34;Author: Tyler K Chafin, Biomathematics and Statistics Scotland&#34;)
        print(&#34;Description: Methods for analysing genetic distances in \
networks.&#34;)
        print(
            &#34;\nMandatory arguments:\n&#34;
            &#34;    -s, --shp       : Path to shapefile containing cleaned, \
contiguous stream reaches\n&#34;
            &#34;                       (can also support geodatabase or GPKG \
files)\n&#34;
            &#34;    -i, --input     : Input .tsv file containing sample \
coordinates\n&#34;
            &#34;    -v, --vcf       : Input VCF file containing genotypes\n\n&#34;

            &#34;General options:\n&#34;
            &#34;    -o, --out       : Output prefix [default=\&#34;out\&#34;]\n&#34;
            &#34;    -O, --gdf_out   : Output driver for annotated geodataframe \
(options \&#34;SHP\&#34;, \&#34;GPKG\&#34;, \&#34;GDB\&#34;)\n&#34;
            &#34;    -C, --concat    : Concatenate all SNPs (\&#34;all\&#34;), by locus \
(\&#34;loc\&#34;), or not at all (\&#34;none\&#34;)\n&#34;
            &#34;    -n, --network   : Provide an already optimized network \
output from a previous run\n&#34;
            &#34;    --overwrite     : Overwrite an input network (Only relevant \
with --network)\n&#34;
            &#34;    -h, --help      : Displays help menu\n&#34;
            &#34;    -r, --run       : Run which steps? Options:\n&#34;
            &#34;        ALL         : Run all steps\n&#34;
            &#34;        GENDIST     : Only calculate genetic distance matrix\n&#34;
            &#34;        STREAMDIST  : Only compute pairwise stream distances\n&#34;
            &#34;        DISTANCES   : Only compute GENDIST + STREAMDIST\n&#34;
            &#34;        IBD         : GENDIST + STREAMDIST + Mantel test\n&#34;
            &#34;        STREAMTREE  : GENDIST + STREAMDIST + fit StreamTree \
model\n&#34;
            &#34;        RUNLOCI     : Run STREAMTREE fitting on each locus\n&#34;
            &#34;    -p, --pop       : Pool individuals based on an input \
population map tsv file\n&#34;
            &#34;        NOTE: The location will be taken as the centroid among \
individual samples\n&#34;
            &#34;    -g, --geopop    : Pool individuals having identical \
coordinates\n&#34;
            &#34;    -c, --clusterpop: Use DBSCAN algorithm to automatically \
cluster populations\n&#34;
            &#34;    --reachid_col   : Attribute name representing primary key in \
shapefile [default=\&#34;HYRIV_ID\&#34;]\n&#34;
            &#34;    --length_col    : Attribute name giving length in kilometers \
[default=\&#34;LENGTH_KM\&#34;]\n\n&#34;
            &#34;    --seed          : Seed for RNG\n\n&#34;

            &#34;Genetic distance options:\n&#34;
            &#34;    -d, --dist      : Use which metric of distance? Options:\n&#34;
            &#34;        Individual-based:\n&#34;
            &#34;          PDIST     : Uncorrected p-distances [# Differences / \
Length]\n&#34;
            &#34;        Frequency models (when using --pop):\n&#34;
            &#34;          FST       : Weir and Cockerham&#39;s Fst formulation \
(=THETAst)\n&#34;
            &#34;          LINFST    : [default] Rousset&#39;s (1997) Fst \
[=Fst/(1-Fst)]\n&#34;
            &#34;          JOST      : Global estimate Jost&#39;s (2008) D\n&#34;
            &#34;          HARMD     : Harmonic mean of Jost&#39;s D per-locus\n&#34;
            &#34;          CHORD     : Cavalli-Sforza and Edwards (1967) chord \
distance\n&#34;
            &#34;          --NOTE: Individual-based metrics can also be computed \
for\n&#34;
            &#34;                    populations. You can set how these are \
aggregated w/ --pop_agg\n&#34;
            &#34;          --NOTE: Multiple loci for PDIST\n&#34;
            &#34;                  will be reported using the method defined in \
--loc_agg\n&#34;
            &#34;    -G, --genmat    : Skip calculation and use the provided \
labeled .tsv matrix\n&#34;
            &#34;    --coercemat     : [Boolean] Coerce negative values in input \
matrix to zero\n&#34;
            &#34;    --het           : [Boolean] Count partial differences [e.g. \
ind1=T, ind2=W]\n&#34;
            &#34;    --global_het    : Estimate Ht using global frequencies \
(default is averaged over pops)\n\n&#34;

            &#34;DBSCAN options (only when --clusterpop):\n&#34;
            &#34;    --min_samples   : Minimum samples per cluster [default=1]\n&#34;
            &#34;    --epsilon       : Maximum distance (in km) within a cluster \
[default=20]\n\n&#34;

            &#34;Aggregation options:\n&#34;
            &#34;    -P, --pop_agg   : Define aggregator function for certain \
genetic distances in pop samples\n&#34;
            &#34;    -L, --loc_agg   : Define aggregator function for aggregating \
locus-wise distances\n&#34;
            &#34;        All of these can take the following options:\n&#34;
            &#34;          ARITH     : [default] Use arithmetic mean\n&#34;
            &#34;          MEDIAN    : Use median distance\n&#34;
            &#34;          HARM      : Use harmonic mean\n&#34;
            &#34;          ADJHARM   : Adjusted harmonic mean (see docs)\n&#34;
            &#34;          GEOM      : Use geometric mean\n&#34;
            &#34;          MIN       : Use minimum distance\n&#34;
            &#34;          MAX       : Use maximum distance\n\n&#34;

            &#34;IBD options:\n&#34;
            &#34;    --perm          : Number of permutations for mantel test \
[def=1000]\n&#34;
            &#34;    --and_log       : Also perform IBD steps with log geographic \
distances\n\n&#34;

            &#34;StreamTree options (see Kalinowski et al. 2008) :\n&#34;
            &#34;    -w, --weight    : Desired weighting for least-squares \
fitting:\n&#34;
            &#34;        Options:\n&#34;
            &#34;          FM67      : Fitch and Margoliash (1967) [w = 1/D^2]\n&#34;
            &#34;          BEYER74   : Beyer et al. (1974) weights [w = 1/D]\n&#34;
            &#34;          CSE67     : [default] Cavalli-Sforza &amp; Edwards (1967) \
[w = 1]\n\n&#34;
        )
        print()
        sys.exit()</code></pre>
</details>
<h3>Methods</h3>
<dl>
<dt id="autostreamtree.params.parseArgs.display_help"><code class="name flex">
<span>def <span class="ident">display_help</span></span>(<span>self, message=None)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">    def display_help(self, message=None):
        if message is not None:
            print(&#34;\n&#34;, message)

        print(&#34;\nautostreamtree\n&#34;)
        print(&#34;Author: Tyler K Chafin, Biomathematics and Statistics Scotland&#34;)
        print(&#34;Description: Methods for analysing genetic distances in \
networks.&#34;)
        print(
            &#34;\nMandatory arguments:\n&#34;
            &#34;    -s, --shp       : Path to shapefile containing cleaned, \
contiguous stream reaches\n&#34;
            &#34;                       (can also support geodatabase or GPKG \
files)\n&#34;
            &#34;    -i, --input     : Input .tsv file containing sample \
coordinates\n&#34;
            &#34;    -v, --vcf       : Input VCF file containing genotypes\n\n&#34;

            &#34;General options:\n&#34;
            &#34;    -o, --out       : Output prefix [default=\&#34;out\&#34;]\n&#34;
            &#34;    -O, --gdf_out   : Output driver for annotated geodataframe \
(options \&#34;SHP\&#34;, \&#34;GPKG\&#34;, \&#34;GDB\&#34;)\n&#34;
            &#34;    -C, --concat    : Concatenate all SNPs (\&#34;all\&#34;), by locus \
(\&#34;loc\&#34;), or not at all (\&#34;none\&#34;)\n&#34;
            &#34;    -n, --network   : Provide an already optimized network \
output from a previous run\n&#34;
            &#34;    --overwrite     : Overwrite an input network (Only relevant \
with --network)\n&#34;
            &#34;    -h, --help      : Displays help menu\n&#34;
            &#34;    -r, --run       : Run which steps? Options:\n&#34;
            &#34;        ALL         : Run all steps\n&#34;
            &#34;        GENDIST     : Only calculate genetic distance matrix\n&#34;
            &#34;        STREAMDIST  : Only compute pairwise stream distances\n&#34;
            &#34;        DISTANCES   : Only compute GENDIST + STREAMDIST\n&#34;
            &#34;        IBD         : GENDIST + STREAMDIST + Mantel test\n&#34;
            &#34;        STREAMTREE  : GENDIST + STREAMDIST + fit StreamTree \
model\n&#34;
            &#34;        RUNLOCI     : Run STREAMTREE fitting on each locus\n&#34;
            &#34;    -p, --pop       : Pool individuals based on an input \
population map tsv file\n&#34;
            &#34;        NOTE: The location will be taken as the centroid among \
individual samples\n&#34;
            &#34;    -g, --geopop    : Pool individuals having identical \
coordinates\n&#34;
            &#34;    -c, --clusterpop: Use DBSCAN algorithm to automatically \
cluster populations\n&#34;
            &#34;    --reachid_col   : Attribute name representing primary key in \
shapefile [default=\&#34;HYRIV_ID\&#34;]\n&#34;
            &#34;    --length_col    : Attribute name giving length in kilometers \
[default=\&#34;LENGTH_KM\&#34;]\n\n&#34;
            &#34;    --seed          : Seed for RNG\n\n&#34;

            &#34;Genetic distance options:\n&#34;
            &#34;    -d, --dist      : Use which metric of distance? Options:\n&#34;
            &#34;        Individual-based:\n&#34;
            &#34;          PDIST     : Uncorrected p-distances [# Differences / \
Length]\n&#34;
            &#34;        Frequency models (when using --pop):\n&#34;
            &#34;          FST       : Weir and Cockerham&#39;s Fst formulation \
(=THETAst)\n&#34;
            &#34;          LINFST    : [default] Rousset&#39;s (1997) Fst \
[=Fst/(1-Fst)]\n&#34;
            &#34;          JOST      : Global estimate Jost&#39;s (2008) D\n&#34;
            &#34;          HARMD     : Harmonic mean of Jost&#39;s D per-locus\n&#34;
            &#34;          CHORD     : Cavalli-Sforza and Edwards (1967) chord \
distance\n&#34;
            &#34;          --NOTE: Individual-based metrics can also be computed \
for\n&#34;
            &#34;                    populations. You can set how these are \
aggregated w/ --pop_agg\n&#34;
            &#34;          --NOTE: Multiple loci for PDIST\n&#34;
            &#34;                  will be reported using the method defined in \
--loc_agg\n&#34;
            &#34;    -G, --genmat    : Skip calculation and use the provided \
labeled .tsv matrix\n&#34;
            &#34;    --coercemat     : [Boolean] Coerce negative values in input \
matrix to zero\n&#34;
            &#34;    --het           : [Boolean] Count partial differences [e.g. \
ind1=T, ind2=W]\n&#34;
            &#34;    --global_het    : Estimate Ht using global frequencies \
(default is averaged over pops)\n\n&#34;

            &#34;DBSCAN options (only when --clusterpop):\n&#34;
            &#34;    --min_samples   : Minimum samples per cluster [default=1]\n&#34;
            &#34;    --epsilon       : Maximum distance (in km) within a cluster \
[default=20]\n\n&#34;

            &#34;Aggregation options:\n&#34;
            &#34;    -P, --pop_agg   : Define aggregator function for certain \
genetic distances in pop samples\n&#34;
            &#34;    -L, --loc_agg   : Define aggregator function for aggregating \
locus-wise distances\n&#34;
            &#34;        All of these can take the following options:\n&#34;
            &#34;          ARITH     : [default] Use arithmetic mean\n&#34;
            &#34;          MEDIAN    : Use median distance\n&#34;
            &#34;          HARM      : Use harmonic mean\n&#34;
            &#34;          ADJHARM   : Adjusted harmonic mean (see docs)\n&#34;
            &#34;          GEOM      : Use geometric mean\n&#34;
            &#34;          MIN       : Use minimum distance\n&#34;
            &#34;          MAX       : Use maximum distance\n\n&#34;

            &#34;IBD options:\n&#34;
            &#34;    --perm          : Number of permutations for mantel test \
[def=1000]\n&#34;
            &#34;    --and_log       : Also perform IBD steps with log geographic \
distances\n\n&#34;

            &#34;StreamTree options (see Kalinowski et al. 2008) :\n&#34;
            &#34;    -w, --weight    : Desired weighting for least-squares \
fitting:\n&#34;
            &#34;        Options:\n&#34;
            &#34;          FM67      : Fitch and Margoliash (1967) [w = 1/D^2]\n&#34;
            &#34;          BEYER74   : Beyer et al. (1974) weights [w = 1/D]\n&#34;
            &#34;          CSE67     : [default] Cavalli-Sforza &amp; Edwards (1967) \
[w = 1]\n\n&#34;
        )
        print()
        sys.exit()</code></pre>
</details>
</dd>
</dl>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="autostreamtree" href="index.html">autostreamtree</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="autostreamtree.params.parseArgs" href="#autostreamtree.params.parseArgs">parseArgs</a></code></h4>
<ul class="">
<li><code><a title="autostreamtree.params.parseArgs.display_help" href="#autostreamtree.params.parseArgs.display_help">display_help</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.8.1</a>.</p>
</footer>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
<script>hljs.initHighlightingOnLoad()</script>
</body>
</html>